src/solver/optimisation/constraint_builder.h

```diff
@@ -0,0 +1,206 @@
+      case Variable::ShortTermStorageInjection:
+        return nativeOptimVar.SIM_ShortTermStorage.InjectionVariable[index_];
+      case Variable::ShortTermStorageWithdrawal:
+        return nativeOptimVar.SIM_ShortTermStorage.WithdrawalVariable[index_];
+      case Variable::ShortTermStorageLevel:
+        return nativeOptimVar.SIM_ShortTermStorage.LevelVariable[index_];
+      case Variable::HydProd:
+        return nativeOptimVar.NumeroDeVariablesDeLaProdHyd[index_];
+      case Variable::HydProdDown:
+        return nativeOptimVar.NumeroDeVariablesVariationHydALaBaisse[index_];
+      case Variable::HydProdUp:
+        return nativeOptimVar.NumeroDeVariablesVariationHydALaHausse[index_];
+      case Variable::Pumping:
+        return nativeOptimVar.NumeroDeVariablesDePompage[index_];
+      case Variable::HydroLevel:
+        return nativeOptimVar.NumeroDeVariablesDeNiveau[index_];
+      case Variable::Overflow:
+        return nativeOptimVar.NumeroDeVariablesDeDebordement[index_];
+      case Variable::FinalStorage:
+        return problemeHebdo.NumeroDeVariableStockFinal[index_];
+      case Variable::LayerStorage:
+        return problemeHebdo.NumeroDeVariableDeTrancheDeStock[index_][0]; // TODO FIXMEEE
+      case Variable::PositiveUnsuppliedEnergy:
+        return nativeOptimVar.NumeroDeVariableDefaillancePositive[index_];
+      case Variable::NegativeUnsuppliedEnergy:
+        return nativeOptimVar.NumeroDeVariableDefaillanceNegative[index_];
+      default:
+        return -1;
+    }
+    }
+
+public:
+    ConstraintBuilder& include(Variable var, double coeff, int timeShift = 0, bool wrap = false)
+    {
+      std::vector<double>& Pi = problemeAResoudre.Pi;
+      std::vector<int>& Colonne = problemeAResoudre.Colonne;
+      int varIndex = getVariableIndex(var, timeShift, wrap);
+      if (varIndex >= 0)
+      {
+        Pi[nombreDeTermes_] = coeff;
+        Colonne[nombreDeTermes_] = varIndex;
+        nombreDeTermes_++;
+      }
+      return *this;
+    }
+    ConstraintBuilder& equal(double rhs)
+    {
+      operator_ = '=';
+      rhs_ = rhs;
+      return *this;
+    }
+    ConstraintBuilder& less(double rhs)
+    {
+      operator_ = '<';
+      rhs_ = rhs;
+      return *this;
+    }
+    ConstraintBuilder& greater(double rhs)
+    {
+      operator_ = '>';
+      rhs_ = rhs;
+      return *this;
+    }
+
+    void build();
+
+private:
+    PROBLEME_ANTARES_A_RESOUDRE& problemeAResoudre;
+    PROBLEME_HEBDO& problemeHebdo;
+    const std::vector<CORRESPONDANCES_DES_VARIABLES>& varNative;
+
+    unsigned int hourInWeek_ = 0;
+    // TODO: vérifier que l'index correspond bien à l'objet concerné (zone, lien, cluster thermique, etc.)
+    unsigned int index_ = 0;
+    char operator_;
+    double rhs_ = 0;
+    int nombreDeTermes_ = 0;
+};
+
+
+class IConstraint {
+public:
+  IConstraint(PROBLEME_HEBDO* problemeHebdo) : problemeHebdo(problemeHebdo), builder(*problemeHebdo, problemeHebdo->CorrespondanceVarNativesVarOptim)
```

Comment:
If we keep the interface, I would suggest to make it pure virtual, to not mix the 2 concerns which are:
- on one side defined the interface contract
- on another side share some code between implementations, which are actually implementation details

See for ex in [CPP core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c121-if-a-base-class-is-used-as-an-interface-make-it-a-pure-abstract-class)

Reply:
Indeed ! Let's this class concrete.